packages/datetime/src/dateRangeInput.tsx

```diff
@@ -461,7 +469,46 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
     // Callbacks - Input
     // =================
 
-    // Key down
+    // instantiate these two functions once so we don't have to for each callback on each render.
+
+    private handleStartInputEvent = (e: React.SyntheticEvent<HTMLInputElement>) => {
+        this.handleInputEvent(e, DateRangeBoundary.START);
+    }
+
+    private handleEndInputEvent = (e: React.SyntheticEvent<HTMLInputElement>) => {
+        this.handleInputEvent(e, DateRangeBoundary.END);
+    }
+
+    private handleInputEvent = (e: React.SyntheticEvent<HTMLInputElement>, boundary: DateRangeBoundary) => {
+        switch (e.type) {
+            case "blur": this.handleInputBlur(e, boundary); break;
+            case "change": this.handleInputChange(e, boundary); break;
+            case "click": this.handleInputClick(e as React.MouseEvent<HTMLInputElement>); break;
+            case "focus": this.handleInputFocus(e, boundary); break;
+            case "keydown": this.handleInputKeyDown(e as React.KeyboardEvent<HTMLInputElement>); break;
+            case "mousedown": this.handleInputMouseDown(); break;
+            default: break;
+        }
+
+        const inputProps = this.getInputProps(boundary);
+        const callbackFn = this.getInputGroupCallbackForEvent(e, inputProps);
+
+        Utils.safeInvoke(callbackFn, e);
+    }
+
+    private getInputGroupCallbackForEvent = (e: React.SyntheticEvent<HTMLInputElement>,
```

Comment:
I put it here because it's so relevant to the top-level input callbacks, but I like it down in the `get*` section too. I'll move it there.